From abc7993f33418bf59f955b37aa8a9c89939d76e1 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michal=20=C4=8Ciha=C5=99?= <michal@cihar.com>
Date: Fri, 22 Nov 2019 13:22:21 +0100
Subject: [PATCH 1/1] Accounts: Fix third party authentication with duplicate
 username
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

The python-social-auth handles this gracefully by generating unique
usernames in such case, but we used the original name for the
verification.

Fixes #3230

Signed-off-by: Michal Čihař <michal@cihar.com>
---
 weblate/accounts/pipeline.py | 12 ++++++------
 1 file changed, 6 insertions(+), 6 deletions(-)

diff --git a/weblate/accounts/pipeline.py b/weblate/accounts/pipeline.py
index a80c9338e5..bc8f4a2688 100644
--- a/weblate/accounts/pipeline.py
+++ b/weblate/accounts/pipeline.py
@@ -249,15 +249,15 @@ def store_params(strategy, user, **kwargs):
     }
 
 
-def verify_username(strategy, backend, details, user=None, **kwargs):
+def verify_username(strategy, backend, details, username, user=None, **kwargs):
     """Verified whether username is still free.
 
     It can happen that user has registered several times or other user has
     taken the username meanwhile.
     """
-    if user or 'username' not in details:
+    if user or not username:
         return
-    if User.objects.filter(username__iexact=details['username']).exists():
+    if User.objects.filter(username__iexact=username).exists():
         raise AuthAlreadyAssociated(backend, 'Username exists')
     return
 
@@ -375,7 +375,7 @@ def notify_connect(strategy, backend, user, social, new_association=False,
         )
 
 
-def user_full_name(strategy, details, user=None, **kwargs):
+def user_full_name(strategy, details, username, user=None, **kwargs):
     """Update user full name using data from provider."""
     if user and not user.full_name:
         full_name = details.get('fullname', '').strip()
@@ -391,8 +391,8 @@ def user_full_name(strategy, details, user=None, **kwargs):
             else:
                 full_name = last_name
 
-        if not full_name and 'username' in details:
-            full_name = details['username']
+        if not full_name and username:
+            full_name = username
 
         if not full_name and user.username:
             full_name = user.username
-- 
2.24.0

